### PR TITLE
Get distance function

### PR DESCRIPTION
**Pull request by GuyAv46**
Added `getDistanceFrom(VecSimIndex *index, size_t id, const void *blob)` to get the distance between the blob to the vector with label `id`

### CODE REVIEW

File: src/VecSim/utils/vec_utils.cpp [hunk 1]
```diff
@@ -47,9 +47,9 @@ int cmpVecSimQueryResultByScore(const VecSimQueryResult *res1, const VecSimQuery
 }
 
 void float_vector_normalize(float *x, size_t dim) {
-    float sum = 0;
+    double sum = 0;
```
Comment (DvirDukhan) [inline src/VecSim/utils/vec_utils.cpp:50]: why the change?

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp [hunk 2]
```diff
@@ -81,6 +81,17 @@ VecSimResolveCode HNSWIndex::resolveParams(VecSimRawParam *rparams, int paramNum
     return (VecSimResolveCode)VecSim_OK;
 }
 
+double HNSWIndex::getDistanceFrom(size_t label, const void *vector_data) {
+    if (this->metric == VecSimMetric_Cosine) {
+        // TODO: need more generic
+        float normalized_data[this->dim];
+        memcpy(normalized_data, vector_data, this->dim * sizeof(float));
+        float_vector_normalize(normalized_data, this->dim);
+        return this->hnsw->getDistanceByLabelFromPoint(label, normalized_data);
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp:90]: please be consistent with the implementation of FLAT and HNSW

File: src/VecSim/vec_sim.h [hunk 3]
```diff
@@ -43,6 +43,18 @@ int VecSimIndex_AddVector(VecSimIndex *index, const void *blob, size_t id);
...
+ * @brief Calculate the distance of a vector from an index to a vector.
+ * @param index the index from which the first vector is located, and that defines the distance
+ * metric.
+ * @param id the id of the vector in the index.
+ * @param blob binary representation of the second vector. Blob size should match the index data
+ * type and dimension.
+ * @return The distance (according to the index's distance metric) between `blob` and the vector
+ * with id `id`.
+ */
+double VecSimIndex_GetDistanceFrom(VecSimIndex *index, size_t id, const void *blob);
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim.h:56]: why double and not float?

File: tests/unit/test_bruteforce.cpp [hunk 4]
```diff
@@ -871,3 +872,61 @@ TEST_F(BruteForceTest, brute_force_resolve_params) {
...
+    // VecSimMetric_IP
+    distances = {-1.4542929723017879e+26, 6.8783602560985513e+29, 3.4384528836109137e+29,
+                 648750694400};
+    for (size_t i = 0; i < n; i++) {
+        dist = VecSimIndex_GetDistanceFrom(index[VecSimMetric_IP], i + 1, query);
+        ASSERT_FLOAT_EQ(dist, (float)distances[i]);
+    }
+
+    // VecSimMetric_Cosine
+    distances = {0, 2, 1.7069573402404785, 2};
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce.cpp:916]: pi is orthogonal to `e` and `gr`?

File: tests/unit/test_bruteforce.cpp [hunk 5]
```diff
@@ -871,3 +872,61 @@ TEST_F(BruteForceTest, brute_force_resolve_params) {
...
+        .algo = VecSimAlgo_BF,
+        .hnswParams = HNSWParams{.type = VecSimType_FLOAT32, .dim = dim, .initialCapacity = n}};
+
+    for (size_t i = 0; i < numIndex; i++) {
+        params.bfParams.metric = (VecSimMetric)i;
+        index[i] = VecSimIndex_New(&params);
+        VecSimIndex_AddVector(index[i], "\100\055\370\124\100\055\370\124", 1); //  pi, pi
+        VecSimIndex_AddVector(index[i], "\100\111\017\333\100\111\017\333", 2); //   e,  e
+        VecSimIndex_AddVector(index[i], "\100\055\370\124\100\111\017\333", 3); //  pi,  e
+        VecSimIndex_AddVector(index[i], "\277\317\033\275\077\317\033\275", 4); // -gr, gr
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce.cpp:893]: gr?

File: src/VecSim/vec_sim.h (hunk 3 above)
Comment (GuyAv46) [reply src/VecSim/vec_sim.h:56]: to be more generic for when we FLOAT64 support

File: tests/unit/test_bruteforce.cpp (hunk 5 above)
Comment (GuyAv46) [reply tests/unit/test_bruteforce.cpp:893]: golden ratio :)

File: tests/unit/test_bruteforce.cpp (hunk 4 above)
Comment (GuyAv46) [reply tests/unit/test_bruteforce.cpp:916]: good catch. (pi, pi) is orthogonal to (-gr, gr) but when normalizing the vectors, normal(pi, pi) == normal(e, e). Ill get to the bottom of it

File: src/VecSim/utils/vec_utils.cpp (hunk 1 above)
Comment (GuyAv46) [reply src/VecSim/utils/vec_utils.cpp:50]: when the floats in the vector are big, the sum is quickly becomes `inf`, and we turn the vector to \overline{0}
this way we can get more precise with the calculation of the norm

**Review by alonre24 — APPROVED** [review]
Looks good! We can also add this function to our benchmarks (in a separate PR)

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp (hunk 2 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp:90]: what do you mean? to have the normalized vector outside the `if`?

File: src/VecSim/spaces/IP/IP.cpp [hunk 6]
```diff
@@ -3,11 +3,15 @@
...
     for (unsigned i = 0; i < qty; i++) {
         res += ((float *)pVect1)[i] * ((float *)pVect2)[i];
     }
-    return 1.0f - res;
+    return res;
 }
+
+float InnerProduct(const void *pVect1, const void *pVect2, const void *qty_ptr) {
+    return 1.0f - InnerProduct_impl(pVect1, pVect2, qty_ptr);
+}
```
Comment (DvirDukhan) [inline src/VecSim/spaces/IP/IP.cpp:17]: empty line

File: src/VecSim/spaces/IP/IP.h [hunk 7]
```diff
@@ -1,3 +1,5 @@
 #pragma once
 
 float InnerProduct(const void *pVect1, const void *pVect2, const void *qty_ptr);
+
+float InnerProduct_impl(const void *pVect1, const void *pVect2, const void *qty_ptr);
```
Comment (DvirDukhan) [inline src/VecSim/spaces/IP/IP.h:5]: empty line

File: tests/unit/test_bruteforce.cpp [hunk 8]
```diff
@@ -871,3 +872,65 @@ TEST_F(BruteForceTest, brute_force_resolve_params) {
...
+        VecSimIndex_AddVector(index[i], v3, 3);
+        VecSimIndex_AddVector(index[i], v4, 4);
+        ASSERT_EQ(VecSimIndex_IndexSize(index[i]), 4);
+    }
+
+    void *query = v1;
+    double dist;
+
+    // VecSimMetric_L2
+    distances = {0, 0.3583844006061554, 0.1791922003030777, 23.739208221435547};
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce.cpp:906]: haven't we talked about calling the space itself?

Comment (GuyAv46) [reply tests/unit/test_bruteforce.cpp:906]: yes, but now we know that the results will be consistent across different hardware and optimizations.
I can change that but Ill have to externalize the metric function

**Review by DvirDukhan — APPROVED** [review]
